### PR TITLE
low efficiency xcorr

### PR DESCRIPTION
**Pull request by trishorts**

### CODE REVIEW

File: EngineLayer/MetaMorpheusEngine.cs [hunk 1]
```diff
@@ -58,6 +57,88 @@ public static double CalculatePeptideScore(MsDataScan thisScan, List<MatchedFrag
...
+                    
+
+                }
+            }
+
+            for (int i = 0; i < scanIntensityArray.Length; i++)
+            {
+                if (Double.IsNaN(scanIntensityArray[i])|| Double.IsNaN(theoreticalFragmentIonIntensityArray[i])   || Double.IsInfinity(scanIntensityArray[i]) || Double.IsInfinity(theoreticalFragmentIonIntensityArray[i]))
+                {
+                    int bubba = 1;
```
Comment (rmillikin) [inline EngineLayer/MetaMorpheusEngine.cs]: 👍

**Comment by acesnik** [comment]
image

File: TaskLayer/MetaMorpheusTask.cs [hunk 2]
```diff
@@ -161,7 +161,12 @@ public static IEnumerable<Ms2ScanWithSpecificMass> GetMs2Scans(MsDataFile myMSDa
                         }
 
                         scansWithPrecursors[i] = new List<Ms2ScanWithSpecificMass>();
-                        IsotopicEnvelope[] neutralExperimentalFragments = Ms2ScanWithSpecificMass.GetNeutralExperimentalFragments(ms2scan, commonParameters);
+                        IsotopicEnvelope[] neutralExperimentalFragments = null;// Ms2ScanWithSpecificMass.GetNeutralExperimentalFragments(ms2scan, commonParameters);
```
Comment (rmillikin) [inline TaskLayer/MetaMorpheusTask.cs]: delete comment

File: TaskLayer/MetaMorpheusTask.cs [hunk 3]
```diff
@@ -248,8 +253,8 @@ public MyTaskResults RunTask(string output_folder, List<DbForTask> currentProtei
             FinishedWritingFile(tomlFileName, new List<string> { displayName });
 
             MetaMorpheusEngine.FinishedSingleEngineHandler += SingleEngineHandlerInTask;
-            try
-            {
+            //try
```
Comment (rmillikin) [inline TaskLayer/MetaMorpheusTask.cs]: uncomment

File: TaskLayer/MyFileManager.cs [hunk 4]
```diff
@@ -95,6 +103,20 @@ public MsDataFile LoadFile(string origDataFile, int? topNpeaks, double? minRatio
...
 #endif
                 }
+
+                if (commonParameters.DissociationType == DissociationType.LowCID)
+                {
+                    // call xcorr filtering on each ms2 scan
+
+                    foreach (var scan in MyMsDataFiles[origDataFile].GetAllScansList().Where(p => p.MsnOrder > 1))
+                    {
+                        var yArray = scan.MassSpectrum.YArray;
```
Comment (rmillikin) [inline TaskLayer/MyFileManager.cs]: delete 113-114?

File: Test/SearchEngineTests.cs [hunk 5]
```diff
@@ -57,6 +57,74 @@ public static void TestClassicSearchEngine()
             Assert.AreEqual("QQQ", allPsmsArray[0].BaseSequence);
         }
 
+        [Test]
+        public static void TestClassicSearchEngineXcorr()
```
Comment (rmillikin) [inline Test/SearchEngineTests.cs]: need some more unit tests. coverage is pretty low (~50%)

**Comment by rmillikin** [comment]
I'm a little worried at how much is changing here and all the if-else stuff. we're almost making two different search programs. duplicated code (or just functionality in general) worries me because of difficulty of maintenance and bugfixes. let's talk about it tomorrow.

File: EngineLayer/ClassicSearch/ClassicSearchEngine.cs [hunk 6]
```diff
@@ -50,14 +49,17 @@ protected override MetaMorpheusEngineResults RunSpecific()
...
             if (Proteins.Any())
             {
-                int maxThreadsPerFile = commonParameters.MaxThreadsToUsePerFile;
-                int[] threads = Enumerable.Range(0, maxThreadsPerFile).ToArray();
-                Parallel.ForEach(threads, (i) =>
+                Parallel.ForEach(Partitioner.Create(0, Proteins.Count), new ParallelOptions { MaxDegreeOfParallelism = commonParameters.MaxThreadsToUsePerFile }, (partitionRange, loopState) =>
+                //Parallel.ForEach(Partitioner.Create(0, Proteins.Count), new ParallelOptions { MaxDegreeOfParallelism = 1 }, (partitionRange, loopState) =>
                 {
-                    for (; i < Proteins.Count; i += maxThreadsPerFile)
+                    for (int i = partitionRange.Item1; i < partitionRange.Item2; i++)
```
[the remark this answers is not in the document]
Comment (trishorts) [reply EngineLayer/ClassicSearch/ClassicSearchEngine.cs]: done

File: EngineLayer/ClassicSearch/ClassicSearchEngine.cs [hunk 7]
```diff
@@ -66,9 +68,11 @@ protected override MetaMorpheusEngineResults RunSpecific()
 
                             foreach (ScanWithIndexAndNotchInfo scan in GetAcceptableScans(peptide.MonoisotopicMass, SearchMode))
                             {
+                                double thisScore = 0;
```
[the remark this answers is not in the document]
Comment (trishorts) [reply EngineLayer/ClassicSearch/ClassicSearchEngine.cs]: i think done...

File: EngineLayer/Indexing/IndexingEngine.cs [hunk 8]
```diff
@@ -121,13 +121,21 @@ protected override MetaMorpheusEngineResults RunSpecific()
             oldPercentProgress = 0;
             for (int peptideId = 0; peptideId < peptidesSortedByMass.Count; peptideId++)
             {
+                //var fragmentMasses = peptidesSortedByMass[peptideId].Fragment(commonParameters.DissociationType, commonParameters.DigestionParams.FragmentationTerminus).Where(x => x.ProductType == ProductType.b || x.ProductType == ProductType.y).Select(m => m.NeutralMass).ToList();
```
[the remark this answers is not in the document]
Comment (trishorts) [reply EngineLayer/Indexing/IndexingEngine.cs]: done

File: TaskLayer/MyFileManager.cs (hunk 4 above)
Comment (trishorts) [reply TaskLayer/MyFileManager.cs]: done

File: EngineLayer/MetaMorpheusEngine.cs [hunk 9]
```diff
@@ -39,18 +38,37 @@ protected MetaMorpheusEngine(CommonParameters commonParameters, List<string> nes
         public static event EventHandler<StringEventArgs> WarnHandler;
 
         public static event EventHandler<ProgressEventArgs> OutProgressHandler;
-        
-        public static double CalculatePeptideScore(MsDataScan thisScan, List<MatchedFragmentIon> matchedFragmentIons, double maximumMassThatFragmentIonScoreIsDoubled)
+
+        public static double CalculatePeptideScore(MsDataScan thisScan, List<MatchedFragmentIon> matchedFragmentIons)
```
Comment (rmillikin) [inline EngineLayer/MetaMorpheusEngine.cs]: I think @zrolfs is using maximumMassThatFragmentIonScoreIsDoubled?

File: EngineLayer/ModernSearch/ModernSearchEngine.cs [hunk 10]
```diff
@@ -64,6 +66,8 @@ protected override MetaMorpheusEngineResults RunSpecific()
                     // get fragment bins for this scan
                     List<int> allBinsToSearch = GetBinsToSearch(scan);
 
+                    //outbut bins AND intensities
```
Comment (rmillikin) [inline EngineLayer/ModernSearch/ModernSearchEngine.cs]: remove comment

File: EngineLayer/ModernSearch/ModernSearchEngine.cs [hunk 11]
```diff
@@ -74,7 +80,13 @@ protected override MetaMorpheusEngineResults RunSpecific()
                     double highestMassPeptideToLookFor = notches.Max(p => p.AllowedInterval.Maximum);
 
                     // first-pass scoring
-                    IndexedScoring(allBinsToSearch, scoringTable, byteScoreCutoff, idsOfPeptidesPossiblyObserved, scan.PrecursorMass, lowestMassPeptideToLookFor, highestMassPeptideToLookFor, PeptideIndex, MassDiffAcceptor, MaxMassThatFragmentIonScoreIsDoubled);
+                    IndexedScoring(allBinsToSearch, intensitiesOfBins, scoringTable, byteScoreCutoff, idsOfPeptidesPossiblyObserved, scan.PrecursorMass, lowestMassPeptideToLookFor, highestMassPeptideToLookFor, PeptideIndex, MassDiffAcceptor, MaxMassThatFragmentIonScoreIsDoubled, commonParameters.DissociationType);
+
+                    //Not necessary
```
[the remark this answers is not in the document]
Comment (rmillikin) [reply EngineLayer/ModernSearch/ModernSearchEngine.cs]: still need to remove comments

File: EngineLayer/ModernSearch/ModernSearchEngine.cs [hunk 12]
```diff
@@ -227,7 +275,14 @@ protected List<int> GetBinsToSearch(Ms2ScanWithSpecificMass scan)
         protected static int BinarySearchBinForPrecursorIndex(List<int> peptideIdsInThisBin, double peptideMassToLookFor, List<PeptideWithSetModifications> peptideIndex)
         {
             int m = 0;
+
             int l = 0;
+            int peptidesInBin = 0;
+            if (peptideIdsInThisBin.Any())
+            {
```
Comment (rmillikin) [inline EngineLayer/ModernSearch/ModernSearchEngine.cs]: I don't think you need any of this? peptideIdsInThisBin should never be empty; if it's not null it should have at least one peptide in it

**Comment by rmillikin** [comment]
looking pretty good

File: EngineLayer/ClassicSearch/ClassicSearchEngine.cs [hunk 13]
```diff
@@ -66,9 +64,10 @@ protected override MetaMorpheusEngineResults RunSpecific()
 
                             foreach (ScanWithIndexAndNotchInfo scan in GetAcceptableScans(peptide.MonoisotopicMass, SearchMode))
                             {
+
                                 List<MatchedFragmentIon> matchedIons = MatchFragmentIons(scan.TheScan, peptideTheorProducts, commonParameters);
 
-                                double thisScore = CalculatePeptideScore(scan.TheScan.TheScan, matchedIons, 0);
+                                double thisScore = CalculatePeptideScore(scan.TheScan.TheScan, matchedIons);
```
[the remark this answers is not in the document]
Comment (trishorts) [reply EngineLayer/ClassicSearch/ClassicSearchEngine.cs]: this no longer has the third field. 
image

File: EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs [hunk 14]
```diff
@@ -163,11 +163,13 @@ private CrosslinkSpectralMatch FindCrosslinkedPeptide(Ms2ScanWithSpecificMass th
                 {
                     List<Product> products = bestPeptide.Fragment(commonParameters.DissociationType, FragmentationTerminus.Both).ToList();
                     var matchedFragmentIons = MatchFragmentIons(theScan, products, commonParameters);
-                    double score = CalculatePeptideScore(theScan.TheScan, matchedFragmentIons, 0);
+                    double score = CalculatePeptideScore(theScan.TheScan, matchedFragmentIons);
```
[the remark this answers is not in the document]
Comment (trishorts) [reply EngineLayer/CrosslinkSearch/CrosslinkSearchEngine.cs]: same

File: EngineLayer/ModernSearch/ModernSearchEngine.cs (hunk 10 above)
Comment (trishorts) [reply EngineLayer/ModernSearch/ModernSearchEngine.cs]: done

File: EngineLayer/ModernSearch/ModernSearchEngine.cs (hunk 12 above)
Comment (trishorts) [reply EngineLayer/ModernSearch/ModernSearchEngine.cs]: done

**Pull request merged**